### PR TITLE
Display traceback as a tree in unhandled errors

### PR DESCRIPTION
**Pull request by romainfrancois**
### Intent

adresses #2393

### Approach

If `rlang` is available, defer to `rlang::trace_back()` and its `format()` method. 

### Automated Tests

> Indicate whether this change includes unit tests or integration tests, or link a work item or pull request to add those tests to another repo. If the change cannot or will not be covered by a test, indicate why.

### QA Notes

> Add additional information for QA on how to validate the change, paying special attention to the level of risk, adjacent areas that could be affected by the change, and any important contextual information not present in the linked issues. 

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

### CODE REVIEW

**Comment by romainfrancois** [comment]
This is work in progress, and currently not handling `rlang::abort()` errors. 

image

For now, just using the text that would be `print()`ed for the `rlang_trace` object, in smaller font, with a red gutter on the left . 

Do we also mean to change the ui in the `Environment > Traceback` pane ? i.e. should this be a tree as well ?

image

**Comment by romainfrancois** [comment]
It's probably a bit fragile at the moment, but I got it to display rlang errors too: 

image

image

To make this happen, I had to work around rlang's workaround: rlang prints errors in the standard output (not standard error) , and so the bit in `consoleWriteError()` that retained elements was not hit. 

The workaround I have in place consists of: 
 - identifying in `.rs.recordTraceback()` if the error is an `rlang` fallback error. This is done heuristically, and @lionel- might have a better way
 - then on the other side, when it is known that this is an rlang error, replace elements added by the last `consoleWriteOutput()`

Unfortunately, this means that `consoleWriteOutput()` always keeps the last results around, i.e. 

```java
      // virtualConsole_.submit(text, className, isError, ariaLiveAnnounce);
      virtualConsole_.submit(text, className, true, ariaLiveAnnounce);
```

which is not great. Maybe instead we could mask `print.rlang_error` somehow ? 

```r
> rlang:::print.rlang_error
function (x, ...) 
{
    writeLines(format(x, ...))
    invisible(x)
}
<bytecode: 0x118fac9b8>
<environment: namespace:rlang>
```

**Comment by romainfrancois** [comment]
Another way perhaps to deal with this would be to rely on custom escape code that would tell the ide that what's coming is special, and would be ignored/not generated in other settings, using e.g. Application Program Command described in https://en.wikipedia.org/wiki/ANSI_escape_code

This would we wouldn't have to print the error on standard output as we do now and then rely on `options(error)` or global handlers etc ... after the fact. All would "just" be handled by the printing as all the information would be available in the string to output. 

cc @gaborcsardi

**Comment by romainfrancois** [comment]
e.g. `cnd_message()` would include some special markers based on e.g. a `RSTUDIO_APC` environment variable or something.

**Comment by gaborcsardi** [comment]
@romainfrancois How is RStudio detecting an error currently? I don't anything in the callback at https://cran.r-project.org/doc/manuals/R-exts.html#Setting-R-callbacks that would tell you that an error happened.

**Comment by lionel-** [comment]
@romainfrancois Good idea. The tagging wouldn't be performed by `cnd_message()` but by `abort()` though, since the former can be called anytime by users. Seeing the tags in output would signal that a unhandled error has occurred and that it was saved in `last_error()`.

RStudio doesn't pipeline user inputs right? It sends R inputs one by one? If it pipelines inputs, this will not work robustly because last-error is transient. In this case, running backtrace capture in error hooks is probably the only viable approach.

**Comment by romainfrancois** [comment]
@gaborcsardi it's not. What may be registered as `options(error)` :  

```r
> getOption("error")
(function()
{
   # When this handler is invoked for an unhandled error happening at
   # the top level, there are four calls on the stack:
   # 1. The traceback recorder
   # 2. The anonymous error handler (set via options below)
   # 3. The error invoker (e.g. stop)
   # 4. The function from which the error was raised
   # So we want there to be at least 5 calls on the stack--otherwise the error
   # is likely to be top-level.
   .rs.recordTraceback(FALSE, 5, .rs.enqueueError)
})()
```

is triggered by the `stop(fallback)` in `signal_abort()` : 

```r
> rlang:::signal_abort
function (cnd, file = NULL) 
{
    .__signal_frame__. <- TRUE
    if (is_true(peek_option("rlang::::force_unhandled_error"))) {
        fallback <- cnd
    }
    else {
        signalCondition(cnd)
        fallback <- cnd
        class(fallback) <- c("rlang_error", "condition")
        fallback$message <- ""
        fallback$rlang$internal$entraced <- TRUE
    }
    poke_last_error(cnd)
    cnd <- cnd_set_backtrace_on_error(cnd, peek_backtrace_on_error())
    msg <- cnd_message(cnd, inherit = TRUE, prefix = TRUE)
    cat_line(msg, file = file %||% default_message_file())
    local_options(show.error.messages = FALSE)
    stop(fallback)
}
<bytecode: 0x150f9fbd8>
<environment: namespace:rlang>
```

which after the fact hunts for an error that was already printed, removes and replaces the dom elements by the fancier display.

**Comment by gaborcsardi** [comment]
Got it. In that case using custom ANSI sequences or markers would make sense indeed.

**Comment by lionel-** [comment]
oh so you'd inspect the output for a tag from an error hook as usual, that makes sense. Please disregard my comment about pipelined input.

**Comment by romainfrancois** [comment]
Thanks to you both. I'll try to advance on that in both `rlang` and here early next week then.

**Comment by romainfrancois** [comment]
see also: https://github.com/r-lib/cli/issues/288

**Comment by romainfrancois** [comment]
For the time being, `consoleWriteOutput()` detects if the output is an rlang error heuristically: 

```java
output.startsWith("\033[1m\033[33mError\033[39m");
```

and the trace is dealt with on the R side. This requires no change in `rlang` but is a bit fragile I'd say. 
This still uses the mechanism of `enqueueError` called from `option(error)`.

**Comment by gaborcsardi** [comment]
Yeah, I agree that this seems a bit fragile.

**Comment by romainfrancois** [comment]
We don't necessarily want to print rlang errors the same way as regular unhandled errors btw, because they have specificities that could be taken advantage: The error title (the `!` bullet), the detail bullets, parent errors (caused by ...).

**Comment by gaborcsardi** [comment]
> perhaps we should also include the traceback metadata directly

Maybe, but `RSTUDIO_APC` will be inherited in subprocesses, so either we need to output a sequence that is harmless in subprocesses, or detect the RStudio console. In this case checking `.Platform$GUI == "RStudio"` might be enough. (Surprisingly this is set to `RGui` at startup time, but for this use case we can probably ignore that.)

**Comment by romainfrancois** [comment]
I'm trying to go back to the initial goal and have the traceback displayed as a tree by working from the structure of the `rlang_trace` object. 

Do we have examples of using `Tree` / `TreeItem` ? I've not been lucky so far: 

image

which is not an improvement over just printing it from R: 

image

**Comment by kevinushey** [comment]
We use a Tree for the document outline:

Screen Shot 2022-03-25 at 5 10 19 PM

But I agree rlang's default printed value is better. Maybe we should just use it as-is?

Also, does rlang support printing of sourcerefs for errors where those are known? Can RStudio navigate to those?

**Comment by romainfrancois** [comment]
Thanks I'll have a look. 

It might just be a case of adding some `style_hyperlink()` on the R side. What I think we can't do is be able to fold branches, but maybe we can do without.

**Comment by romainfrancois** [comment]
Actually there are 3 choices + hide, based on `format.rlang_trace(simplify=)`:  

image

**Comment by romainfrancois** [comment]
Current status. With this code: 

```r
library(rlang)

x1 <- function() x2()
x2 <- function() tryCatch(x3())
x3 <- function() tryCatch(x4())
x4 <- function() { y <- 2; abort(c("Something wrong", i = "A hint about it")) }

# x1()
```

and 

image

we get tree / list / hide: 

image

image

image

The "Rerun with debug" gets to the right frame: 

image

but due to the way `abort()` works, we still get the print of the error, which is a bit misleading: 

image


Some issues: 

  - to make this work, I'm using a modified `signal_abort()` that adds a special marker when `signal_abort()` prints the error (now based on an env var, but should be something else): 

```r
  msg <- cnd_message(cnd, inherit = TRUE, prefix = TRUE)

  if (identical(Sys.getenv("RSTUDIO_APC"), "true")) {
    msg[1L] <- paste0("\u001B_rlang_error\u0007", msg[1L])
  }
```

because ShellWidget only records dom elements that are added on its `consoleWriteError()` method; and so marking it allows to also capture on `consoleWriteOutput()` with the special marker: 

```java
   @Override
   public void consoleWriteOutput(String output)
   {
      clearPendingInput();

      // identify if the output is an rlang error.
      Match apcMatch = AnsiCode.APC_ESCAPE_PATTERN.match(output, 0);
      boolean isError = false;
      if (apcMatch != null && StringUtil.equals("rlang_error", apcMatch.getGroup(1)))
      {
         isError = true;

         // remove the APC signal
         output = output.substring(apcMatch.getValue().length());
      }
      
      output(output, styles_.output(), isError, false /*ignoreLineCount*/,
            isAnnouncementEnabled(AriaLiveService.CONSOLE_LOG));
   }
```

The point is that new dom elements are only recorded when needed (for performance reasons). 

```java
      // Pick up the elements emitted to the console by this call. If we get
      // extended information for this error, we'll need to swap out the simple
      // error elements for the extended error element.
      if (isError)
         recordErrorElements(text);
```



 - the tracebacks generated by rlang don't at the moment link, as was the case with the custom code. 

 - I have to verify what happens when rlang is not available ... 

(so some more work here).

**Comment by romainfrancois** [comment]
Currently (no longer needing a change in `rlang`.

image

image

when `rlang` is not available, we only get `list` and this uses the previous implementation: 

image

Pending: need to find a way to have both rlang nicer display and the links.

**Comment by romainfrancois** [comment]
At this point, I guess we could either improve how `rlang:::format.rlang_trace()` works to add hyperlinks, which I guess would be an `rlang` pull request or pick up the rlang trace here and re-generate the tree here. This potentially would give more freedom, but not sure it's worth it.

File: src/cpp/session/modules/SessionErrors.R [hunk 1]
```diff
@@ -103,9 +103,39 @@
...
+         if (length(calls) > 2) {
+            call <- calls[[length(calls) - 3L]]
+            is_abort <- is.language(call) && identical(as.character(call[[1]]), "signal_abort")
+         }
+         
+         if (is_abort) {
+            # retrieve information from the real condition, instead of the fallback error
+            cnd <- frames[[length(frames) - 3L]]$cnd # using rlang::last_error() does not give the footer
+            trace <- cnd$trace
+            message <- paste0(paste(rlang:::cnd_message(cnd, inherit = TRUE, prefix = TRUE), collapse = "\n"), "\n")
```
Comment (lionel-) [inline src/cpp/session/modules/SessionErrors.R:121]:
```suggestion
            message <- paste0(paste(rlang::cnd_message(cnd, inherit = TRUE, prefix = TRUE), collapse = "\n"), "\n")
```

File: src/cpp/session/modules/SessionErrors.R [hunk 2]
```diff
@@ -103,9 +103,39 @@
...
+         } else {
+            # a regular stop() error, we can keep the message, but still
+            # retrieve the traceback with rlang
+            trace <- rlang::trace_back(bottom = frames[[length(frames) - 2L]])
+         }
+         
+         trace <- c(
+            paste(rlang:::format.rlang_trace(trace, simplify = "none"), collapse = "\n"), 
+            paste(rlang:::format.rlang_trace(trace, simplify = "branch"), collapse = "\n")
```
Comment (lionel-) [inline src/cpp/session/modules/SessionErrors.R:130]:
```suggestion
            paste(format(trace, simplify = "none"), collapse = "\n"), 
            paste(format(trace, simplify = "branch"), collapse = "\n")
```

**Review by lionel- — COMMENTED** [review]
I'd be happy to add links to `rlang:::format.rlang_trace()`. Please let me know if more input is needed from me.

**Comment by romainfrancois** [comment]
Thanks @lionel- . If we add links on the `rlang` side then this part of the pr is done. 

@kevinushey this is still wip though, I need to work more on:
 - cases where `rlang` is not available
 - I need to understand better the modification of the stack wrt python
 - how it looks

**Comment by romainfrancois** [comment]
Current status: 

```r
library(rlang)

x1 <- function() x2()
x2 <- function() tryCatch(x3())
x3 <- function() tryCatch(x4())
x4 <- function() { 
  y <- 2; 
  abort(c("Something wrong", i = "A hint about it"))
}

y1 <- function() y2()
y2 <- function() tryCatch(y3())
y3 <- function() tryCatch(y4())
y4 <- function() { 
  y <- 2; 
  stop("something wrong")
}
```

If `rlang` is installed, we get to choose between a tree or a list representation of the traceback:

image

image

(the file location are not yet clickable, but this is a feature request we can tackle on the rlang side). 


Same when error does not originate from `rlang::abort()`, i.e. we still use `rlang` to help with the traceback: 

image

image

image

If rlang is not installed, we get the previous behavior (no tree):

image

But the file positions can be clicked. 

That's a bit unfortunate, and we really should have the same rendering in both cases. 

Also, when `rlang` is available, this completely bypasses the whole `stack <- ` and `ammended_stack <-` code in `.rs.recordTraceback()`.

**Pull request closed without merging**